Spawn to deserialise blocks offline

martinsumner opened the pull request:
Hypothesis is that the growth in the heap necessary due to continual term_to_binary calls to deserialise blocks is wasting memory - so do this memory-intensive task in a short-lived process.

martinsumner committed:
Hibernate on range query

If Block Index Cache is not full, and we're not yielding

martinsumner committed:
Spawn to deserialise blocks offline

Hypothesis is that the growth in the heap necessary due to continual term_to_binary calls to deserialise blocks is wasting memory - so do this memory-intensive task in a short-lived process.

ThomasArts commented on src/leveled_sst.erl:
```diff
@@ -2164,8 +2164,11 @@ check_blocks([Pos|Rest], BlockPointer, BlockLengths, PosBinLength,
+    Pid =
+        spawn(
+            ?MODULE, check_block, [self(), BlockPos, BlockBin, PressMethod]),
+    Result = receive {checked_block, Pid, R} -> R end,
```
This is just code to verify an idea... otherwise, we should have used some more caution here on timeout of a possible crash of spawned code, etc.... but not for this experiment.

martinsumner committed:
Start with hibernate_after option

martinsumner committed:
Always build BIC

Testing indicates that the BIC itself is not a primary memory issue - the primary issue is due to a lack of garbage collection and a growing heap.

This change enhances the patch to offline serialisation so that:
- get_sqn & get_kv are standardised to build the BIC, and hibernate when it is built.
- the offline PId is linked to crash this process on failure (as would happen now).

martinsumner committed:
Merge branch 'mas-i1820-hibernateafter' into mas-i1820-offlinedeserialisation

martinsumner committed:
Merge branch 'mas-i1820-hibernateonkvr' into mas-i1820-offlinedeserialisation

martinsumner committed:
Standardise spawning for check_block/3

Now deserialise in both parts of the code.
